sound/soc/sof/ipc4-mtrace.c

```diff
@@ -424,9 +426,19 @@ static int ipc4_mtrace_enable(struct snd_sof_dev *sdev)
 	msg.primary |= SOF_IPC4_MOD_INSTANCE(SOF_IPC4_MOD_INIT_BASEFW_INSTANCE_ID);
 	msg.extension = SOF_IPC4_MOD_EXT_MSG_PARAM_ID(SOF_IPC4_FW_PARAM_SYSTEM_TIME);
 
-	/* The system time is in usec, UTC, epoch is 1601-01-01 00:00:00 */
-	kt = ktime_add_us(ktime_get_real(), FW_EPOCH_DELTA * USEC_PER_SEC);
-	system_time = ktime_to_us(kt);
+	/* do calibration to get the latency beween timestamp generation and fw processing */
+	send_time = local_clock();
+	msg.data_size = sizeof(system_time);
+	msg.data_ptr = &system_time;
+	/* get system time from fw */
+	ret = iops->set_get_data(sdev, &msg, msg.data_size, false);
+	if (ret)
+		return ret;
+	reply_time = local_clock();
+	fw_receive_latency = (reply_time - send_time) / 2;
```

Comment:
this is measuring the IPC latency, not the time difference. I don't understand how this can be used for calibration.

Reply:
sure, this is for IPC latency, time difference is calculated in FW which compares host time stamp and dsp time stamp to get the difference, then fw log will add this difference for eah log time.

the latency is added to current host time to minimize the error between host time and dsp time 
system_time =local_clock() + fw_receive_latency;

Reply:
@plbossart @RanderWang @ujfalusi I wonder if it would be better to live with an error in the timestamps. The SOF_IPC4_FW_PARAM_SYSTEM_TIME IPC4 is a very crude tool to set FW time and no claim beyond this have been made. There are no provisions in the spec to accurately synchronize the time, so the time it takes for host time to be queried (t=T0), IPC to be sent, received and processed at FW (t=T4), is not covered and by the time FW sets a new time, that timestamp is already old.

I understand @RanderWang you are trying to use a IPC send-response transaction to estimate the time delta between T0 and T4, so that effectively the time FW ends up setting as system time, will be closer to kernel time.

This is a very pragmatic approach, but it of course has limits. Given we cannot ensure accuracy of the timestamp, I wonder if it would be better to not try to remove the error but simply send "local_clock()" here. There will be error of hundreds of microseconds or more, but at least the timestamps are aligned within tens of milliseconds. And the code is probably easier to follow and maintain (in both kernel and FW).

If one needs more accurate timestamp synchronization between kernel and DSP events, we have have separate hardware tools for this (e.g. https://www.kernel.org/doc/html/v6.1/trace/intel_th.html for Intel). Not enabled for SOF yet, but that's just matter of work. Or we enable a proper time sync between DSP and host in SW/FW with new IPC extensions.

To me, a sub-100ms accuracy is enough for debug convenience purposes. If we need more, I think a great deal more effort needs to be put into this.

Reply:
> Given we cannot ensure accuracy of the timestamp, I wonder if it would be better to not try to remove the error but simply send "local_clock()" here. There will be error of hundreds of microseconds or more, but at least the timestamps are aligned within tens of milliseconds. And the code is probably easier to follow and maintain (in both kernel and FW).

Amen. It will be wrong anyway, so at least keep the code simple and predictable and the error always on the same side. Nothing worse than a random error that will be sometimes early and sometime late.

> To me, a sub-100ms accuracy is enough for debug convenience purposes. If we need more, I think a great deal more effort needs to be put into this.

That's it. Either it should be _really_ done properly with great accuracy, or it should not be done at all.